dask/array/core.py

```diff
@@ -1109,12 +1109,20 @@ def itemsize(self):
         """ Length of one array element in bytes """
         return self.dtype.itemsize
 
+    @property
+    def name(self):
+        return self._name
+
+    @name.setter
+    def name(self, val):
+        self._name = val
+        # Clear the key cache when the name is reset
+        self._cached_keys = None
+
     def _keys(self, *args):
         if not args:
-            try:
+            if self._cached_keys is not None:
```

Comment:
Do you know `__slots__` affects dictionary key sharing? https://www.python.org/dev/peps/pep-0412/

Maybe now that you check of `self._cached_keys is not None` you can add `self._cached_keys = None` to the `__init__`.